### ISSUE TITLE
If you want to...

### ISSUE THREAD

**Issue by jimmywarting** [michaelrhodes/piece-length#2]
This was a fun challenge trying to exclude the dependency of `closest-to` while making it smaller and more performant

```js
module.exports = function bytes(bytes, p) {
  bytes /= 1024
  for (
    p = 20; 
    13 < p && bytes < 2 << p; 
    p--
  );
  return 2 << p + (bytes >= ((2 << p) + (2 << p + 1)) / 2)
}
```

```js
module.exports=function(b,a){b/=1024;for(a=20;13<a&&b<2<<a;a--);return 2<<a+(b>=((2<<a)+(2<<a+1))/2)}
```

```js
module.exports=(b,a)=>{b/=1024;for(a=20;13<a&&b<2<<a;a--);return 2<<a+(b>=((2<<a)+(2<<a+1))/2)}
```

Tried everything i could to get rid of the `return` keyword and just return a value with arrow function.<br> To do that i could not use any `for` or `while` so i changed it to a recursive function but it ended up being longer (and worse)

**Comment by michaelrhodes** [comment]
I _do_ want to, and I (finally) have!

Released as `1.0.1` and cited you as an author in `package.json`

Thanks for the awesome contribution :)

**Comment by jimmywarting** [comment]
Woho, i bearly remember how i manage to do it or how it even works. bitwise is hard.

**Comment by jimmywarting** [comment]
Aha, i have beaten myself! I found easier way and a better bitwise operation to figure out if something is closer and then afterwards round it!

take `x=1024` and `y=2048`
you want to know if `val=1333` is closer to x or y
how do you do compare that? you take the middle value of x and y and compare 1024+512 < val (whats is the comparison in binary?) `11000000000`
```
 |-512
 |
11000000000 = 1536
|
|-1024
```
so you see, having the piece work with two inital ones make it easier to compare if something lower or larger, but you don't want to return the second `1` so how do you get rid of it?
you shift it one step to the left so you get a much larger number

```
011000000000  (<< 1)
110000000000 = 3072
```
then you use the and-operator and you get all bits that are equal `p & p << 1`
```
011000000000  (&)
110000000000
---------------
010000000000 === 1024
```

so here it is, a much simpler function:
```js
function length (b, p = 3 << 13) { // p starts of with two initial 11
  b /= 1024
  for(;b > p; p <<= 1); // shift p to the left while bytes is still bigger
  return p & p << 1 // remove the half comparison bit
}
```


and a recursive just for fun, since it's now easer to use it without a return due to fewer operations
```js
let l=(b,p=3<<13)=>b/1024>p?l(b,p<<=1):p&p<<1;module.exports=l
```

...when i tried to understand my own code i started to understand more deeply what was happening

**Comment by jimmywarting** [comment]
lol, when i found out what log2 dose you can just do:
```js
module.exports = x => Math.max(8192, 1 << Math.log2(x/1024)+.5|0)
```

can throw in a max as well so it don't become lower than `1 << 13`
ruined all the fun of bitwise hacking :P

**Comment by michaelrhodes** [comment]
Excellent code golfing! Thanks for the breaking it all down too. So I was going to release a new version using the simpler function because it supports IE natively and brotlifies 5 bytes smaller than log2, but then I ran a quick benchmark and discovered not only that the simpler function is significantly slower than `1.0.1`, but that the log2 implementation is _insanely_ fast:

```
1.0.0 x 53,303,088 ops/sec ±0.10% (91 runs sampled)
1.0.1 x 198,805,511 ops/sec ±0.13% (90 runs sampled)
simpler (non-recursive) x 81,454,810 ops/sec ±0.12% (90 runs sampled)
simpler (recursive) x 17,549,539 ops/sec ±0.22% (93 runs sampled)
log2 x 896,762,268 ops/sec ±0.13% (88 runs sampled)
```

Needless to say I’ve published `2.0.0` using the log2 implementation.

For reference, the benchmark code I used:
```js
var benchmark = require('benchmark') // 2.1.4

// Removed require(`${implementation}`) calls for brevity

var bytes = 4.5 * Math.pow(2, 30)

benchmark.Suite()
  .add('1.0.0', function () {
    v100(bytes)
  })
  .add('1.0.1', function () {
    v101(bytes)
  })
  .add('simpler (non-recursive)', function () {
    simpler(bytes)
  })
  .add('simpler (recursive)', function () {
    recursive(bytes)
  })
  .add('log2', function () {
    log2(bytes)
  })
  .on('error', error)
  .on('cycle', cycle)
  .run()

function error (e) {
  console.error(e.target.error.stack)
}

function cycle (e) {
  console.log(String(e.target))
}
```

Thanks again! :)

**Comment by feross** [comment]
Updating to 2.0.0 breaks the tests in `create-torrent`. Are you sure this works the same as the previous version?

This is the failing travis build: https://travis-ci.org/webtorrent/create-torrent/builds/571609546#L297

You can reproduce it like this:

```js
pieceLength(4) // returns 16384 in v1, returns 33554432 in v2
```

I assume this wasn't intentional?

**Comment by michaelrhodes** [comment]
It was not, fixing at the moment 😅

**Comment by feross** [comment]
> It was not, fixing at the moment 😅

No worries. I've been there :)

**Comment by michaelrhodes** [comment]
@feross Fixed in `2.0.1`. I also added a test for a `pieceLength(1)` to help prevent future regressions

**Comment by feross** [comment]
You're awesome! Thanks for the quick fix! 🤩

![giphy](https://user-images.githubusercontent.com/121766/54960514-2af5b300-4f1a-11e9-8dc0-3a5378ed0aac.gif)
